---
title: "Kochia_QTL_mapping_workshop"
output: word_document
---

## Extracting genotype information from a vcf file

Here, we take a vcf file that has been filtered, and extract information from the genotype field.

```{r}
#install.packages('vcfR')
library(vcfR)

#read in an actual vcf file and convert/save as a vcfR object
vcf <- read.vcfR( "Kochia_F2_GBS.vcf", verbose = FALSE)

#extract the genotypes from the file
head(vcf[,1:5])
geno <- extract.gt(vcf)
geno_df <- as.data.frame(geno, sep = ',')
head(geno_df[,1:5])
write.csv(geno_df, file = "Kochia_genotypes.csv")

#Assign alleles based on parental genotypes by calling an external python script
#Arguments IN ORDER: /path/to/python_script.py /path/to/geno_df.csv Name_of_R_parent Name_of_S_parent /path/to/alleles_out.csv
system('python assign_alleles.py ./Kochia_genotypes.csv R_parent S_parent ./Kochia_alleles.csv', wait = FALSE)
```

## Loading information into a "cross" object using the qtl2 package

```{r}
library(qtl2)
#load in the data by calling the control file
cross <- read_cross2("kochia_qtl_control_file.yaml")

#plot a histogram of biomass
hist(cross$pheno, xlab = "Biomass (g)", main = "Histogram of F2 biomass")
```

##Prep the data for a QTL scan and conduct the scan

```{r}
#calculate genetic probabilities at each marker. Default error rate is .0001
pr <- calc_genoprob(cross=cross, map=cross$gmap, error_prob = .002)

#look at the names of the entries in the pr data structure
names(pr)

#look at the dimension names for the chromosome 6 array
dimnames(pr$'chr6')

#view the first three rows of genotype probabilities for a genotyped marker on chromosome 6
(pr$`chr6`)[1:3,,"chr6_52000000"]

#plot the probabilities of individual 1 genotypes across chromosome 6
plot_genoprob(pr, map=cross$gmap, ind = 1, chr = "chr6")

#haley-knott regression plots genotype probability on a spectrum (x-axis) and phenotype on the y-axis. A linear regression (including any covariates you provide) will show if there is an effect of genotype on phenotype.
#perform genome scan using haley-knott regression on the data previously generated
out <- scan1(genoprobs=pr, pheno =cross$pheno)

#plot the LOD scores for the liver LOD scores
plot_scan1(out, map = cross$gmap, lodcolumn = "biomass")
```

## Using permutation to determine a significance threshold for QTL peaks

```{r}
#permutation mixes the phenotypes up so there should be no relationship between genotype and phenotype
#the 95th percentile calculated LOD score from these permutation is considered the 95% confidence threshold for statistical significance 
operm <- scan1perm(genoprobs = pr, pheno = cross$pheno, n_perm = 1000)

#this outputs a list of the maximum LOD score achieved from each permutation
#plot these values
{hist(operm[,'biomass'], breaks = 50, xlab = "LOD", main = "LOD scores for biomass scan with threshold in red")
abline(v = summary(operm)[,'biomass'], col = 'red', lwd = 2)}

#operm has calculated the 95th percentile (5% of random LOD are above this line)
#return thresholds of different alphas
summary(operm, alpha= c(.02, .05))

#to find significant peaks, we use the findpeaks function with the scan output, thresholds, and marker map
#save the permutation data so we can call it later
thr <- summary(operm)

#find the peaks with 95% confidence (thr) and find the genomic interval that is 95% likely to contain a causative locus (based on a bayesian credible interval)
find_peaks(scan1_output = out, map = cross$gmap, threshold = thr, prob = 0.95, expand2markers = FALSE)

#find multiple peaks on a chromosome with 'peakdrop' to tell how far the LOD needs to drop between peaks
find_peaks(scan1_output = out, map = cross$gmap, threshold = thr, peakdrop = 1.8, prob = 0.95, expand2markers = FALSE)
```

## Extracting predicted genes in your QTL intervals

```{bash}
#this bash command will filter only predicted genes in your chr1 interval and extract the predicted function of the protein product (all from the gff annotation file)
#in this interval we find many proteins, but the Glutathione S-transferase-like protein is an interesting candidate for sure
grep gene Kochia_reduced_annotations.gff3 | awk '$1=="chr1"' | awk '$4>81000000' | awk '$4<93000000' > ./chr1_qtl_genes.txt

grep gene Kochia_reduced_annotations.gff3 | awk '$1=="chr1"' | awk '$4>81000000' | awk '$4<93000000' | cut -d$'\t' -f 9 | rev | cut -d '=' -f 1 | rev | sort | uniq > ./chr1_qtl_gene_names.txt

head ./chr1_qtl_gene_names.txt
wc -l ./chr1_qtl_gene_names.txt
grep Cytochrome ./chr6_qtl_gene_names.txt
```

```{bash}
#this will do the same for the chr6 interval
#notice the Cytochrome P450 protein in the interval. We will have to validate to be sure it is responsible for resistance.
grep gene Kochia_reduced_annotations.gff3 | awk '$1=="chr6"' | awk '$4>2500000' | awk '$4<3500000' > ./chr6_qtl_genes.txt

grep gene Kochia_reduced_annotations.gff3 | awk '$1=="chr6"' | awk '$4>2500000' | awk '$4<3500000' | cut -d$'\t' -f 9 | rev | cut -d '=' -f 1 | rev | sort | uniq > ./chr6_qtl_gene_names.txt

head ./chr6_qtl_gene_names.txt
wc -l ./chr6_qtl_gene_names.txt
grep Cytochrome ./chr6_qtl_gene_names.txt
```

